README
Time::Interval
==============

This package installs the Time::Interval perl module.
Time::Interval's primary purpose is to take a given number
of seconds and return the number of days, hours, minutes,
and seconds are represented. This is handier than you 
think, and of course, I'm fairly certain that I'm not the
the first/best/only one to have done this with perl.
However, this is my attempt, and I think at the very least
the name makes the intent of the module clear.

thank you and enjoy,

-Andrew ;-)


Changes
Revision history for Perl extension Time::Interval.

1.232		Thu Jan 31 21:41:40 CST 2013
		those damn ._* files os x kindly inserted into the distribution
		I have removed them. sorry everyone.

1.231	        Wed Jan 30 07:58:21 CST 2013
		forgot to add LICENSE to manifest.
		incremented version for CPAN upload

1.23		Tue Jan 29 10:13:54 CST 2013
		updated LICENSE.
		this perl module is released under the same license
		as perl itself

1.22		Wed May 30 22:57:32 EDT 2007
		applied patch to fix another warning when using -w flag.
		thanks, Sean <sean.robinson@sccmail.maricopa.edu>

1.21		Mon Jan 15 16:01:00 EST 2007
		applied patch to fix warnings when using perl -w flag.
		thanks, Ton <ton.voon@altinity.com>

1.2		Tue Dec 12 08:46:00 EST 2006
		added 'short' option to getInterval, and applied 
		patch from Alberto Simones adding 'Short' option to parseInterval
		which generates a string with abbreviated output in dhms format
		(see pod)
		Thanks, Alberto!
		
1.1		Thu Nov  2 18:08:48 GMT 2006
		don't convert input strings to epoch if already converted 

1.0.5		Mon May 17 18:18:26 EDT 2004
		strange array problems fixed.

1.0.4		Mon May 17 17:42:24 EDT 2004
		fixed bug in coalesce causing undef array values.

1.0.3		Thu May 13 17:24:58 EDT 2004
		added coalesce method, takes a series of time intervals and returns 
		the minimum number of intervals based on overlapping timespans.

1.0.2		Tue Oct 14 11:00:05 EDT 2003
		submitted by: Doug Ferguson <dougf@bookings.nl>.
		Removed useage of 'our' to declare class global variables,
		instead, using the 'use vars ()' method. In addition to being
		'more correct', this also allows older versions of perl to 
		use Time::Interval. Thanx, Doug!

1.0.1		Mon Apr 28 14:25:58 2003
		fixed dependancy list in Makefile.PL, added Date::Parse

1.0		Fri Apr 25 10:58:30 2003
		initial revision for release to CPAN



test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

#########################

# change 'tests => 1' to 'tests => last_test_to_print';

use Test;
BEGIN { plan tests => 12 };
use Time::Interval;
ok(1); # If we made it this far, we're ok.

#########################

#test getInterval
my $str1 = "1/15/03 12:34:32 EDT 2003";
my $str2 = "4/25/03 11:24:00 EDT 2003";
print "testing getInterval between:\n\t$str1\n\t$str2\n";
if (my $data = getInterval($str1,$str2)){
	foreach (keys %{$data}){ print "[$_]: $data->{$_}\n"; }
	ok(1);
}else{
	ok(0);
}

#test convertInterval
my %data = (
	'days'		=> 70,
	'hours'		=> 16,
	'minutes'	=> 56,
	'seconds'	=> 18
);

print "testing convertInterval on:\n";
foreach (keys %data){ print "[$_]: $data{$_}\n"; }
foreach ('days','hours','minutes','seconds'){
	my $num = convertInterval(
		ConvertTo	=> $_,
		%data
	) || ok(0);
	print "converting to $_ ...: $num\n";
	ok(1);
}

#test parseInterval
print "testing parseInterval (data):\n";
foreach ('days','hours','minutes','seconds'){
	print "12345 $_ is: ...\n";
	my $string = parseInterval(
		$_		=> 12345,
		String	=> 1
	) || ok(0);
	print "\t$string\n";
	ok(1);
}

#test parseInterval
print "testing parseInterval w/abbreviated string (data):\n";
foreach ('days','hours','minutes','seconds'){
	print "12345 $_ is: ...\n";
	my $string = parseInterval(
		$_		=> 12345,
		Small	=> 1
	) || ok(0);
	print "\t$string\n";
	ok(1);
}

#test coalesce
print "testing coalesce on: \n";
my @date_ranges  = (
#	[ "5/13/04 10:00:00 EDT", "5/13/04 11:00:00 EDT" ],
#	[ "5/13/04 13:00:00 EDT", "5/13/04 14:00:00 EDT" ],
#	[ "5/13/04 10:50:00 EDT", "5/13/04 12:00:00 EDT" ],
#	[ "5/10/04 10:50:00 EDT", "5/11/04 12:00:00 EDT" ],
#	[ "5/13/04 9:50:00 EDT",  "5/13/04 11:00:00 EDT" ],
	[ "20040429 11:34:16", "20040430 02:19:30" ],
	[ "20040429 11:34:16", "20040501 02:19:31" ],
	[ "20040429 11:34:16", "20040502 02:19:31" ],
	[ "20040429 11:34:16", "20040503 02:19:31" ],
	[ "20040429 11:34:16", "20040504 02:19:31" ],
	[ "20040429 11:34:16", "20040504 14:04:58" ]

);
foreach (@date_ranges){ print $_->[0], " - ", $_->[1], "\n"; }

$tmp = coalesce(\@date_ranges);
print "coalesced dates:\n";
foreach (@{$tmp}){
	print $_->[0], " - ", $_->[1], "\n";
}
ok(1);

Interval.pm
###################################################
## Interval.pm		(Time::Interval)
## Andrew N. Hicox	<andrew@hicox.com>
## http://www.hicox.com
##
## a module for dealing with time intervals
###################################################


## Global Stuff ###################################
package	Time::Interval;
use		strict;
require	Exporter;

#class global vars ...
use vars qw($VERSION @EXPORT @ISA %intervals);
@ISA 		= qw(Exporter);
@EXPORT		= qw(&parseInterval &convertInterval &getInterval &coalesce);
$VERSION	= 1.232;
#what everything is worth in seconds
%intervals 	= (
	'days'		=> ((60**2) * 24),
	'hours'		=> (60 **2),
	'minutes'	=> 60,
	'seconds'	=> 1
);


## getInterval ####################################
sub getInterval {
	my $date1 = shift();
	my $date2 = shift();
	my $string = shift() || "";
	if ( (! $date1) || (! $date2) ){
		warn ("two dates are required for the getInterval method");
		return (undef);
	}
	require Date::Parse;
	foreach ($date1, $date2){
		$_ = Date::Parse::str2time($_) || do {
			warn ("failed to parse date: $!\n");
			return (undef);
		};
	}
	
	my %args = ( seconds => abs($date1 - $date2) );
	if ($string =~/^small/i){
		$args{'Small'} = 1;
	}elsif($string !~/^\s*$/){
		$args{'String'} = 1;
	}
	my $data = parseInterval(%args);
	return ($data);
}


## convertInterval ################################
#'days'		=> $num,
#'hours'	=> $num,
#'minutes'	=> $num,
#'seconds'	=> $num,
#'ConvertTo'	=> 'days'|'hours'|'minutes'|'seconds'
sub convertInterval {
	my %p = @_;
	#ConvertTo, default is seconds
	exists($p{'ConvertTo'}) || do {
		$p{'ConvertTo'} = "seconds";
		warn ("convertInterval: using default ConvertTo (seconds)") if $p{'Debug'};
	};
	#convert everything to seconds
	my $seconds = 0;
	foreach ("days","hours","minutes","seconds"){
		if (exists($p{$_})){ $seconds += ($intervals{$_} * $p{$_}); }
	}
	#send it back out into the desired output
	return (($seconds/$intervals{$p{'ConvertTo'}}));
}


## parseInterval ##################################
#'days'		=> $num,
#'hours'	=> $num,
#'minutes'	=> $num,
#'seconds'	=> $num,
sub parseInterval {
	my %p = @_;
	#convert everything to seconds
	my $seconds = convertInterval(%p);
	#do the thang
	my %time = (
		'days'		=> 0,
		'hours'		=> 0,
		'minutes'	=> 0,
		'seconds'	=> 0
	);
	while ($seconds > 0){
		foreach ("days","hours","minutes","seconds"){
			if ($seconds >= $intervals{$_}){
				$time{$_} ++;
				$seconds  -= $intervals{$_};
				last;
			}
		}
	}
	#return data
        if ($p{'Small'} && $p{'Small'} != 0) {
		#return a string?
		my @temp = ();
		foreach ("days","hours","minutes","seconds"){
			if ($time{$_} > 0){
				push (@temp, "$time{$_}".substr($_,0,1));
			}
		}
        return join (" ", @temp) || "0s";
	}elsif ($p{'String'} && $p{'String'} != 0){
		#return a string?
		my @temp = ();
		foreach ("days","hours","minutes","seconds"){
			if ($time{$_} > 0){
                          if ($time{$_} == 1) {
				push (@temp, "$time{$_} ".substr($_,0,-1));
                              } else {
				push (@temp, "$time{$_} $_");
                              }
			}
		}
		return (join (", ", @temp)) || "0 seconds";
	}else{
		#return a data structure
		return (\%time);
	}	
}


## coalesce #######################################
#coalesce([ [$start1, $end1], [$start2, $end2] ... ])
sub coalesce {
	require Date::Parse;
	my $intervals = shift() || [];
	my %epoch_map = ();
	my ($flag, $repeat) = (0,1);
	
	#convert each start / end to an epoch pair and stash 'em in epoch_map
	foreach my $int (@{$intervals}) {
		foreach (@{$int}){
			
			## only convert if it's not already epoch time
			my $epoch = "";
			if ($_ =~/^(\d{10})$/){
				$epoch = $1;
			}else{
				my $epoch = Date::Parse::str2time($_);
			}
			$epoch_map{$epoch} = $_;
			$_ = $epoch;
		}
	}

	#sort 'em by start time
	@{$intervals} = sort { $a->[0] <=> $b->[0] } @{$intervals};
	
	#flatten 'em
	while ($repeat == 1) {
		@{$intervals} = sort {
			#if it's not an array ref, it's been destructo'd
			if ( (ref($a) eq "ARRAY") && (ref($b) eq "ARRAY") ){
			
				#if b is inside a
				if (($b->[0] >= $a->[0]) && ($b->[0] <= $a->[1])){
					#if b's end time is greater than a's, update a's end time
					if ($b->[1] > $a->[1]){ $a->[1] = $b->[1]; }
					#destructo b
					$b = ();
					$flag = 1;
					return (1);
				#if a is inside b
				}elsif (($a->[0] >= $b->[0]) && ($a->[0] <= $b->[1])){
					#if a's end time is greater than b's, update b's end time
					if ($a->[1] > $b->[1]){ $b->[1] = $a->[1]; }
					#destructo a
					$a = ();
					$flag = 1;
					return (0);
				}else{
					return (0);
				}
		
			}else{
				return (1);
			}
	
		} @{$intervals};
	
		#weed out null elements
		my $i = 0;
		foreach (@{$intervals}){
			if ( ref($_) ne "ARRAY" ){ splice (@{$intervals}, $i, 1); }
			$i ++;
		}
		
		#decide wether or not to repeat
		if ($flag == 1){ 
			$repeat = 1;  
			$flag = 0;
		}else{ 
			$repeat = 0;
		}
	}
	
	#replace the epoch's with their time string equivalents
	foreach my $pair (@{$intervals}){
		if ( ref($pair) eq "ARRAY"){
			foreach (@{$pair}){ $_ = $epoch_map{$_}; }
		}
	}
	
	#weed out any remaining bum elements
	my $i = 0;
	foreach (@{$intervals}){
		unless (ref($_) eq "ARRAY"){ splice (@{$intervals}, $i, 1); }
		$i ++;
	}
	return ($intervals);
}
